### DISCUSSION TITLE
Set slotduration based on eventduration?

### DISCUSSION THREAD

**Discussion by Azur1337** [vkurko/calendar#230] [Q&A]
Im currently working on a project where I need to create a schooltimetable. This is my current progress.
![image](https://github.com/vkurko/calendar/assets/82388136/ac5dd546-af5b-4dec-9a7a-f3a5afc2501c)

Right now I have the calendars SlotMinTime be the first events starting time and the SlotMaxTime the last events ending time.
However, I want the slots to be the same duration as the events, in this example the first slot should be 07:30-09:00, second slot 09:00-09:15 and the third slot 09:15-10:45.
Ive been trying to find a solution for this for a few hours now and havent been successful yet.
Any ideas?

**Comment by vkurko** [comment]
You have a very specific requirement and unfortunately there is no solution for it out of the box. Perhaps you should reduce the `slotDuration`, for example, to 15 minutes and then somehow hide the extra time labels via CSS, leaving only the necessary ones...

**Reply by Azur1337** [reply]
That could work.
But maybe is there a way to manually set the duration of single slots?

And on another basis: Is there a way to use 24h format instead of 12h?

**Reply by vkurko** [reply]
> is there a way to manually set the duration of single slot?

Unfortunately, you can change the length of all slots at once, rather than each one individually. See [slotDuration](https://github.com/vkurko/calendar#slotduration).

> Is there a way to use 24h format instead of 12h?

See [slotLabelFormat](https://github.com/vkurko/calendar?tab=readme-ov-file#slotlabelformat).

**Reply by Azur1337** [reply]
Thank you for the answer.
However im sorry for yet another question. I cant quite figure out how to parse an array of objects into events.
Im calling an API which response with all lessons 
`["{\n  \"class_id\":3,\n  \"end_time\":\"2024-01-28T23:00:00Z\",\n  \"exam\":false,\n  \"lesson_id\":1,\n  \"name\":\"Englisch\",\n  \"start_time\":\"2024-01-28T23:00:00Z\",\n  \"substitute\":false,\n  \"teacher_id\":10\n}"]`

I want to use end_time as end, start_time as start and name as title but I just cant seem to make it work

**Reply by vkurko** [reply]
Simply convert the received data into the format [required by the calendar](https://github.com/vkurko/calendar?tab=readme-ov-file#parsing-event-from-a-plain-object):

```js
events: receivedData.map(entry => ({
  start: entry.start_time,
  end: entry.end_time,
  title: entry.name
}))
```

**Reply by Azur1337** [reply]
This doesnt seem to work as this is my event array 
`Array [ {…}, {…} ]
​
0: Object { start: "2024-01-28T23:00:00Z", end: "2024-01-28T23:00:00Z", title: "Englisch" }
​​
end: "2024-01-28T23:00:00Z"
​​
start: "2024-01-28T23:00:00Z"
​​
title: "Englisch"
​​
<prototype>: Object { … }
​
1: Object { start: "2024-02-01T23:00:00Z", end: "2024-02-01T23:00:00Z", title: "Deutsch" }
​
length: 2
​
<prototype>: Array []
` 
And nothing shows up. Not quite sure why

**Reply by vkurko** [reply]
> start: "2024-01-28T23:00:00Z", end: "2024-01-28T23:00:00Z"

The duration of your events is zero because the end time is equal to the start time of the event.

**Reply by Azur1337** [reply]
`7: Object { start: "2024-02-06T12:15:16Z", end: "2024-02-06T13:45:16Z", title: "test" }
​​
end: "2024-02-06T13:45:16Z"
​​
start: "2024-02-06T12:15:16Z"
​​
title: "test"`

Our lessons were all stuck at 23:00 so we fixed that but even with the working time, the lessons wont show up in the calendar?

**Reply by vkurko** [reply]
I've prepared a [demo](https://jsbin.com/betozefaxe/edit?html,output) with your events data. Does it work for you?

**Reply by Azur1337** [reply]
Yes, manually adding Events works fine, however I make an Api call which responds with several informations including a start time, end time and a name. I then map these to fit an event objects attributes, but none of the events show up in the calendar, its just empty.

**Reply by vkurko** [reply]
If you can provide a working example of your code like my demo, then I can help you.

**Reply by Azur1337** [reply]
Hello,
Im terribly sorry for such a late reply, I didnt see your request.
[Here](https://jsbin.com/hadaxuzizu/edit?html,output) is an example of my code where I am using a method you mentioned in an earlier reply.
I hope this can give you more context.

**Reply by vkurko** [reply]
After fixing the syntax error in your code, I can see the event in the calendar (on 2024-02-13):

![image](https://github.com/vkurko/calendar/assets/1073174/91e2af5f-1971-4d2a-81d1-81808148bc62)

Syntax error:

instead of

```js
    const ec = new EventCalendar(document.getElementById('ec'), {
        view: 'timeGridWeek',
        date: '2024-02-06',
        events: lessons.map(entry => ({
          start: entry.start_time,
          end: entry.end_time,
          title: entry.name,
    });
```
there should be

```js
    const ec = new EventCalendar(document.getElementById('ec'), {
        view: 'timeGridWeek',
        date: '2024-02-06',
        events: lessons.map(entry => ({
          start: entry.start_time,
          end: entry.end_time,
          title: entry.name,
        }))
    });
```

**Reply by Azur1337** [reply]
Sorry for the confusion. As we are using svelte we're not using the ES6 way of usage. Instead we use 
```js
let plugins = [TimeGrid];
  let options = {
    view: "timeGridWeek",
    firstDay: 1,
    eventBackgroundColor: "#769CD9",
    flexibleSlotTimeLimits: true,
    slotDuration: "00:15",
    slotHeight: 24,
    slotMinTime: "07:30",
    slotMaxTime: "07:30",
    events: lesson.map(entry => ({
          start: entry.start_time,
          end: entry.end_time,
          title: entry.name,
        }))
  };```
  
  And call the calendar via `<Calendar {plugins} {options} />`
  
  For us, the events are still not showing up despite the fixed syntax

**Reply by vkurko** [reply]
These values must not be equal, otherwise there is no time period to display:

```js
slotMinTime: "07:30",
slotMaxTime: "07:30",
```

**Reply by Azur1337** [reply]
Subject: Sincere Apology for the Oversight Regarding Calendar Rendering

Dear vkurko,

I hope this message finds you well. I am writing to extend my heartfelt apologies for the oversight that led to a significant waste of your time. Upon careful review and reflection, it has come to my attention that the error we encountered in rendering the calendar was due to attempting to render it before the associated array was properly filled. This realization brings forth a deep sense of regret and responsibility on my part, and I sincerely apologize for any inconvenience and frustration this may have caused you.

As a fellow developer, I understand the value of time and effort invested in our projects. Your dedication and hard work in crafting the calendar functionality deserve commendation, and it deeply pains me to realize that my oversight resulted in a needless waste of your valuable time. Please accept my sincerest apologies for this oversight.

Moving forward, I am committed to taking proactive measures to prevent similar incidents from occurring in the future. I will ensure thorough testing and validation processes are in place to catch such issues early on and prevent them from affecting our collaborative efforts.

I want to express my gratitude for your patience and professionalism throughout this ordeal. Your willingness to collaborate and troubleshoot the issue together reflects the true spirit of teamwork, and I am genuinely appreciative of your support.

Please know that I hold the utmost respect for your skills and contributions to our project. Your expertise and insights are invaluable assets, and I am committed to fostering a more conducive environment for collaboration and success moving forward.

Once again, I apologize wholeheartedly for the oversight and any inconvenience it may have caused you. Please do not hesitate to reach out if there is anything I can do to make amends or assist you further.

Thank you for your understanding and patience.

Warm regards,

Azur

I am stupid.

**Reply by Azur1337** [reply]
> These values must not be equal, otherwise there is no time period to display:
This does in fact work, as flexibleSlotTimeLimits is true

**Reply by vkurko** [reply]
No problem :smiley:
